### DISCUSSION TITLE
Any solution to callback hell?

### DISCUSSION THREAD

**Discussion by xiewei20082008** [actor-framework/actor-framework#1215] [General]
Hi,

We got a lot of nested requests in our project. I copied several lines of example as below. I know using make_function_view in an actor might cause dead lock, so is there any solution to such callback hell? Thank you in advance!

```
using inc_actor_t = typed_actor<
    replies_to<int>::with<int>
>;

inc_actor_t::behavior_type inc_impl(inc_actor_t::pointer self) {
    return {
        [=](int x) {
            return x+1;
        }
    };
}

void b_impl(event_based_actor *self, inc_actor_t a) {
    self->request(a, caf::infinite, 1).then(
        [=](int x) {
            self->request(a, caf::infinite, x).then(
                [=](int x) {
                    self->request(a, caf::infinite, x).then(
                        [=](int x) {
                            self->request(a, caf::infinite, x).then(
                                [=](int x) {
                                    cout << x << endl;
                                }
                            );
                        }
                    );
                }
            );
        }
    );
}

int main() {
    actor_system_config cfg;
    actor_system sys{cfg};
    inc_actor_t inc_actor = sys.spawn(inc_impl);
    sys.spawn(b_impl, inc_actor);
}
```

**Comment by Neverlord** [comment]
As long as your work flow is a sequence, where you need the output of each step as input to the next step, I don't think there's much CAF could do in this case to provide a better API without something like Monads in the language.

Coroutines may help eventually, but I don't think they're well supported at the moment.

**Comment by xiewei20082008** [comment]
I understand. Thank you for your reply.

**Comment by Neverlord** [comment]
Btw, we're always open for API suggestions should you have an idea how to streamline the code for such use cases. 🙂

**Comment by melg8** [comment]
My personal take on this would be, as suggested by @Neverlord using monadic bindings around types, returned by function view on actor.
Disclaimer: this is not production ready implementation, should be viewed only as code example.
Something like this:
``` C++
#include <iostream>
#include <vector>

#include "caf/all.hpp"
#include "caf/expected.hpp"

using namespace caf;

using inc_actor_t = typed_actor<replies_to<int>::with<int>>;

[[nodiscard]] inc_actor_t::behavior_type inc_impl(inc_actor_t::pointer self) {
  return {[=](int x) { return x + 1; }};
}

template <typename T, typename F>
[[nodiscard]] decltype(auto) mbind(const expected<T>& expected, F&& f) {
  using ReturnType =
    typename std::remove_cv<decltype(f(expected.value()))>::type;

  if (expected) {
    return f(expected.value());
  }
  if constexpr (!std::is_same_v<ReturnType, void>) {
    return ReturnType{expected.error()};
  }
}

template <typename T, typename F>
auto operator|(const expected<T>& expected, F&& f) {
  return mbind(expected, f);
}

void b_impl(event_based_actor* self, const inc_actor_t& a) {
  const auto increment = make_function_view(a);
  increment(1) | increment | increment | increment |
    [=](int x) { std::cout << x << std::endl; };
}

int main() {
  actor_system_config cfg;
  caf::core::init_global_meta_objects();
  actor_system sys{cfg};
  inc_actor_t inc_actor = sys.spawn(inc_impl);
  sys.spawn(b_impl, inc_actor);
}

```

**Comment by Neverlord** [comment]
I'd like to see first-class support for monads in C++ one day. Given that the `std::expected` proposal seems to go nowhere, I doubt we'll ever see it, though. I'd rather not go into this direction with CAF TBH. Once we go beyond small examples, supporting this comes with too many strings attached.

FWIW, you could use CAF's pipelining here instead:

```c++
void b_impl(event_based_actor* self, const inc_actor_t& a) {
  auto increment = make_function_view(a * a * a * a);
  std::cout << increment(1) << std::endl;
}
```

We're using `operator*`, because you cannot overload `.` in C++. The semantics is the same as the `.`-operator of function composition, i.e., `a.b = a(b(...))`.

I do have to admit that the pipelining is terribly documented at the moment, though.

**Comment by xiewei20082008** [comment]
> My personal take on this would be, as suggested by @Neverlord using monadic bindings around types, returned by function view on actor.
> Disclaimer: this is not production ready implementation, should be viewed only as code example.
> Something like this:
> 
> ```c++
> #include <iostream>
> #include <vector>
> 
> #include "caf/all.hpp"
> #include "caf/expected.hpp"
> 
> using namespace caf;
> 
> using inc_actor_t = typed_actor<replies_to<int>::with<int>>;
> 
> [[nodiscard]] inc_actor_t::behavior_type inc_impl(inc_actor_t::pointer self) {
>   return {[=](int x) { return x + 1; }};
> }
> 
> template <typename T, typename F>
> [[nodiscard]] decltype(auto) mbind(const expected<T>& expected, F&& f) {
>   using ReturnType =
>     typename std::remove_cv<decltype(f(expected.value()))>::type;
> 
>   if (expected) {
>     return f(expected.value());
>   }
>   if constexpr (!std::is_same_v<ReturnType, void>) {
>     return ReturnType{expected.error()};
>   }
> }
> 
> template <typename T, typename F>
> auto operator|(const expected<T>& expected, F&& f) {
>   return mbind(expected, f);
> }
> 
> void b_impl(event_based_actor* self, const inc_actor_t& a) {
>   const auto increment = make_function_view(a);
>   increment(1) | increment | increment | increment |
>     [=](int x) { std::cout << x << std::endl; };
> }
> 
> int main() {
>   actor_system_config cfg;
>   caf::core::init_global_meta_objects();
>   actor_system sys{cfg};
>   inc_actor_t inc_actor = sys.spawn(inc_impl);
>   sys.spawn(b_impl, inc_actor);
> }
> ```

Thank you for your reply. As using scoped_actor in an actor may cause a potential deadlock, so I am trying to keep myself avoiding using make_function_view here.

**Comment by xiewei20082008** [comment]
Hi @Neverlord, is that reasonable to add a feature for the request interface with signature:

`request :: Actor, Timeout, Ts ... -> CafFuture<R> `

Then we can do something like this.
```
void b_impl(event_based_actor *self, inc_actor_t a) {
    auto rp = self->make_response_promise<int>();
    CafFuture<int> a= self->request(a, caf::infinite, 1);
    a.flatmap(
        [=](int v) {
            return self->request(v, caf::infinite, 1);
        }
    ).flatmap(
        [=](int v) {
            return self->request(v, caf::infinite, 1);
        }
    ).flatmap(
        [=](int v) {
            return self->request(v, caf::infinite, 1);
        }
    ).map(
        [=](int v) {
            return v+1;
        }
    );
    rp.deliver(a);
}
```

Thank you for your time.

**Reply by Neverlord** [reply]
I'm not sure how one would implement this kind of interface without adding tons of metaprogramming utilities to CAF plus requiring quite a bit additional bookkeeping at the actor. If you think you can implement something like this with clean and maintainable code without causing pessimization for the simple case: feel free to give it a shot. 🙂

**Reply by Neverlord** [reply]
While not quite the same, you can actually get close to your proposal with the new flow API!

```c++
caf::behavior foo(caf::event_based_actor* self, caf::actor worker) {
  return {
    [self, worker](int32_t x) {
      // Handle for delaying our response.
      auto rp = self->make_response_promise<int>();
      // First request with the original value.
      self->request(worker, caf::infinite, x)
        .as_observable<int32_t>()
        // Second request with the first increment.
        .flat_map([self, worker](int32_t v1) {
          return self->request(worker, caf::infinite, v1)
            .as_observable<int32_t>();
        })
        // Third request with the second increment.
        .flat_map([self, worker](int32_t v2) {
          return self->request(worker, caf::infinite, v2)
            .as_observable<int32_t>();
        })
        // Increment the value for good measure.
        .map([](int32_t v3) { return v3 + 1; })
        // Ship response.
        .for_each([rp](int32_t val) mutable { rp.deliver(val); });
      return rp;
    },
  };
}
```

You could also throw in a `do_on_error(...)` if you wish to propagate an error somewhere along the way to the client.

That code should run on `topic/neverlord/flow`. The branch still has some rough edges and lacks documentation, but it's not far off from getting merged back to the main branch for the next release.

**Comment by Neverlord** [comment]
To add another spin to the discussion: I had the idea of using coroutines for this in the back of my head for a while now. The feature did make C++20 and the stackless design is precisely what we'd want for a new CAF API. Without monads in the language, this is probably the "next best thing". I think we could get away with something like this:

```cpp
self->async([=](auto& async) {
  auto v1 = co_await async.request(a, caf::infinite, 1);
  auto v2 = co_await async.request(a, caf::infinite, v1);
  // ...
});
```

On an error, this would call the default error handler on `self` (just like omitting an explicit error handler would). Of course the question is how far we'd like to push this. Rather than doing this `async` trick on the current actors, we could also think about a new actor type instead, built with coroutines from the ground up. CAF used to have blocking actors based on Boost.Context that we ditched eventually because the cost of context switches made them run significantly slower than event-based actors. Since the compiler transforms coroutine code into lightweight, resumable objects, I think a coroutine-based design should not have much performance cost to it.

I couldn't play with it yet since compiler support just isn't there at the moment, but I thought I'd put the idea out there at least.

**Reply by dominiklohmann** [reply]
Did you consider doing this via a  `caf::function_view`-esque API by returning an expected? Some form of `caf::co_function_view` that takes both self and the actor it creates a view for.

```cpp
auto fn = caf::co_function_view{self, a};
auto v1 = co_await fn(caf::infinite, v1);
if (!v1)
  return handle_error(v1.error());
auto v2 = co_await fn(caf::ifinite, *v1);
if (!v1)
  return handle_error(v2.error());
do_something_with(*v2);
```

**Reply by Neverlord** [reply]
That's also a neat idea. 🙂

At least the way I understand coroutines, we'd probably still wrap this into some `self->async(...)` call to get a handle to something that the actor / CAF can schedule:

```c++
behavior foo_actor(event_based_actor* self) {
  return {
    [=](worker_actor a) -> int32_t {
      return self->async([self](auto&) {
        auto fn = caf::co_function_view{self, a};
        auto v1 = co_await fn(caf::infinite, v1);
        if (!v1)
          co_return v1;
        auto v2 = co_await fn(caf::ifinite, *v1);
        co_return v2;
    }
  };
}
```

**Reply by dominiklohmann** [reply]
Or you can dispatch based on the return value, which is probably easier to implement.

```c++
behavior foo_actor(event_based_actor* self) {
  return {
    [=](worker_actor a) -> caf::lazy_result<int32_t> {
      auto fn = caf::co_function_view{self, a};
      auto v1 = co_await fn(caf::infinite, v1);
      if (!v1)
        co_return v1.error();
      co_return fn(caf::infinite, *v1);
    },
  };
}
```

**Comment by xiewei20082008** [comment]
Interesting. Actually, we did some experimental trials to connect the CAF-scheduled world with other scheduled worlds. 

A small attempt we tried with boost asio coroutine:

```cpp
template<class Actor, class R, class... Ts>
void co_impl_with_callback(event_based_actor* self, Actor act, std::function<void (std::optional<R>, caf::error)> f, Ts... xs) {
    self->set_down_handler(
        [=](down_msg &msg) {
            self->quit();
        }
    );
    self->monitor(act);
    self->request(act, caf::infinite, xs...).then(
        [=](R x) {
            f(x, caf::error{});
        },
        [=](caf::error &e) {
            f(nullopt, e);
        }
    );
}

template <class Actor, class... Ts,
        class R =
            caf::function_view_flattened_result_t<
                typename caf::response_type<
                    typename Actor::signatures,
                    caf::detail::implicit_conversions_t<
                    typename std::decay<Ts>::type
                    >...
                >::tuple_type>
>
caf::expected<R> make_co_fv(yield_context yield, const Actor& x, Ts... xs) {
    auto &sys = x.home_system();
    CoroutineProcessExp<R> cp;
    auto f = cp.yield_to_f(yield);
    sys.spawn(co_impl_with_callback<Actor, R, Ts...>, x, f, xs...);
    return cp.result();
}
behavior foo_actor(event_based_actor* self) {
    return {
        [=](worker_actor a) -> int {
            autp rp = self->make_response_promise<int>();
            asio::spawn([](yield_context yield) {
                auto v1 = make_co_fv(yield, a, 1);                        // line 41
                if(!v1)
                    ...;
                auto v2 = make_co_fv(yield, a, *v1);
                if(!v2)
                    ...;
                rp.deliver(*v2);                                                     // line 47
            });
            return rp;
        }
    };
}
```

I know an issue here is that, for line41-47, these lines of code are scheduled by asio workers, so we cannot operate the actor's state here, but this is still acceptable. 
Another issue is that it looks like the response_promise was not designed to be used outside the actor scheduler (It has an `local_actor *` reference in its implementation), so rp has potential crash risk with its lifespan.

Not sure if there is any way for temporarily leveraging third-party coroutine before  CAF has its native implementation.

**Reply by Neverlord** [reply]
> trials to connect the CAF-scheduled world with other scheduled worlds.

Uh... don't do that. This adds intra-actor concurrency and breaks several fundamental assumptions in CAF. Calling it "potential crash risk" is putting it mildly. The response handler you pass to `request(...).then(...)` could be a safe entry point for something like this.

Btw, you can also schedule tasks directly to CAF. There's currently no convenient API for it but just offloading some task for later would look somewhat like this (untested):

```cpp
template <class F>
void run_later(caf::actor_system& sys, F fn) {
  struct impl : caf::resumable {
    F fn_;
    explicit impl(F f) : fn_(std::move(f)) {
      // nop
    }
    resumable::subtype_t subtype() const override {
      return resumable::function_object;
    }
    resume_result resume(execution_unit*, size_t) override {
      fn_();
      return resumable::done;
    }
  };
  sys.scheduler().enqueue(new impl{std::move(fn)});
}

```

Of course ASIO has more than just dispatching tasks later, just pointing it out. 🙂

**Comment by dominiklohmann** [comment]
I implemented a coroutine-enabling wrapper here: tenzir/vast#2043

This fully works, but it's written for CAF 0.17 right now as we're still on that (expecting to upgrade to 0.18 in the next 1-2 months actually).

I would gladly contribute this to upstream CAF as a separate library that can optionally be linked against, e.g., `caf::co`. It has to be a separate library since it only compiles with clang 13+ for now (still have to add support for gcc 10+), so the compiler requirements are much stricter than for CAF itself. Any thoughts, @Neverlord?

**Reply by Neverlord** [reply]
Sounds interesting!

> expecting to upgrade to 0.18 in the next 1-2 months actually

I currently have my hands full with the flow API. Since the code is going to change again in the near future, I'd say I give it a look after it's been upgraded. 🙂

**Comment by xiewei20082008** [comment]
@dominiklohmann For your "coroutine-enabling wrapper", is there a potential coroutine frame leaking, as the coroutine's releasing depends on co_result_promise's deconstruction which might not happen if an actor is killed before the `then scope` execution.

```
auto handle_value = traits::wrap(
      [=, this](typename traits::argument_type value) mutable noexcept {
        response_ = traits::unwrap(std::move(value));
        VAST_ASSERT(coroutine);
        coroutine.resume();   // what if this is never being called.
      });
    auto handle_error = [=, this](caf::error& err) mutable noexcept {
      response_ = std::move(err);
      VAST_ASSERT(coroutine);
      coroutine.resume();
    };
```

It seems like the lifespan of coroutine frame need to be bound to response_promise.

**Reply by dominiklohmann** [reply]
That's a really nice catch, and indeed a bug assuming CAF doesn't call the error handlers for requests if the request-issuing actor was killed. I think this can be solved by wrapping the whole hosting actor, and simply attaching a functor that frees any leftover coroutine frames. I'll have to give this some thought.

The work I've done on that front is currently on hold until we get green light from customers to upgrade to gcc-11, for which we fixed the ICE in gcc upstream that we get with that branch. I'm likely to pick it up again in a few months when that happens.

**Reply by dominiklohmann** [reply]
Just out of curiosity: Are you using this already? I have some further plans with it, so if you're open to chatting/brainstorming about it please reach out to `dominik.lohmann(at)tenzir.com`.

**Comment by xiewei20082008** [comment]
Recently, I've got some time to go back to the project using CAF. I wrote a light-version coroutine interface for those who may be interested.

https://github.com/xiewei20082008/caf_co

Usage example:

```cpp
caf_co<int> inc_once(actor_requester auto *self, actor add_actor, int base) {
	auto res = co_await co_request<int>(self, add_actor, caf::infinite, 1, base);
	cout << "inc_once get res: " << res << endl;
	co_return *res;
}

caf_co<int> inc_twice(actor_requester auto *self, actor add_actor, int base) {
	auto res1 = co_await inc_once(self, add_actor, base).to_awaiter(self);
	auto res2 = co_await inc_once(self, add_actor, *res1).to_awaiter(self);
	co_return *res2;
}


behavior test_impl(event_based_actor *self, actor add_actor) {
	return {
		[=](int x) -> result<int> {
			auto co_env = make_co_env<int>(self);
			return co_env->make_co( [self, x, add_actor](auto co) -> caf_co<int>{
				auto res = co_await inc_twice(self, add_actor, x).to_awaiter(self);
				cout << *res << endl;
				auto res1 = co_await co_request<int>(self, add_actor, caf::infinite, 1, *res);
				co_return *res1;
			});
		}
	};
}
```
